prov/mrail/src/mrail_rma.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2018 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+typedef ssize_t mrail_rma_msg_fn(struct fid_ep *ep_fid,
+		const struct fi_msg_rma *msg, uint64_t flags);
+
+static ssize_t mrail_ep_post_rma(struct fid_ep *ep_fid,
+		const struct fi_msg_rma *msg, uint64_t flags,
+		mrail_rma_msg_fn *rma_fn)
+{
+	struct mrail_ep *mrail_ep;
+	struct mrail_addr_key *mr_map;
+	fi_addr_t *rail_fi_addr;
+	uint32_t rail;
+	ssize_t ret;
+
+	size_t num_rails;
+	size_t remaining_len;
+	size_t chunk_len;
+	size_t offset;
+	size_t sublen;
+	size_t i;
+
+	struct mrail_rma_req *req;
+
+	if (msg->iov_count > 1) {
+		FI_WARN(&mrail_prov, FI_LOG_EP_DATA,
+				"Cannot handle iov_count > 1");
+		return -FI_ENOSYS;
+	}
+
+	if (msg->rma_iov_count > 1) {
+		FI_WARN(&mrail_prov, FI_LOG_EP_DATA,
+				"Cannot handle rma_iov_count > 1");
+		return -FI_ENOSYS;
+	}
+
+	mrail_ep = container_of(ep_fid, struct mrail_ep, util_ep.ep_fid.fid);
+	rail_fi_addr = ofi_av_get_addr(mrail_ep->util_ep.av, (int)msg->addr);
+	mr_map = (struct mrail_addr_key*)msg->rma_iov->key;
+
+	num_rails = mrail_ep->num_eps;
+	remaining_len = msg->msg_iov->iov_len;
+	chunk_len = remaining_len / num_rails;
```

Comment:
Can we guarantee that remaining_len is divisible by num_rails? Otherwise, chunk_len = (remaining_len + num_rails - 1) / num_rails; would be better to prevent having data left over.

Reply:
Ah, might be better to do something like `chunk_len = 1 + ((remaining_len - 1) / num_rails);` to prevent overflow possibility. May need to check if remaining_len is 0, but it shouldn't be.

Reply:
Please correct me if I'm wrong but I *think* that the case where remaining_len is not divisible by num_rails is being handled during the calculation of each subrequest's length later (line 92). chunk_len is the length of all the subrequests but the last one. The last subrequest picks up the remainder.

Of course, this doesn't work if remaining_len < num_rails...

Reply:
You're right, though using the upper ceiling will probably have a slightly better data distribution as well as handle remaining_len < num_rails better.

It looks like if remaining_len < num_rails case the data will still be sent through, it'll just loop through sending 0 for the first n-1 rails and then send the entire thing at the end.